server/util.go

```
@@ -41,43 +46,62 @@ func NewContext(v *viper.Viper, config *tmcfg.Config, logger log.Logger) *Contex
+// the application configuration. Command handlers can fetch the server Context
+// to get the Tendermint configuration or to get access to Viper.
+func InterceptConfigsPreRunHandler(cmd *cobra.Command) error {
+	rootViper := viper.New()
+	rootViper.BindPFlags(cmd.Flags())
+	rootViper.BindPFlags(cmd.PersistentFlags())
+
+	serverCtx := NewDefaultContext()
+	config, err := interceptConfigs(serverCtx, rootViper)
+	if err != nil {
+		return err
+	}
 
-		config, err := interceptConfigs(ctx, rootViper)
-		if err != nil {
-			return err
-		}
+	logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+	logger, err = tmflags.ParseLogLevel(config.LogLevel, logger, tmcfg.DefaultLogLevel())
+	if err != nil {
+		return err
+	}
 
-		logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
-		logger, err = tmflags.ParseLogLevel(config.LogLevel, logger, tmcfg.DefaultLogLevel())
-		if err != nil {
-			return err
-		}
+	if rootViper.GetBool(tmcli.TraceFlag) {
+		logger = log.NewTracingLogger(logger)
+	}
 
-		if rootViper.GetBool(tmcli.TraceFlag) {
-			logger = log.NewTracingLogger(logger)
-		}
+	serverCtx.Config = config
+	serverCtx.Logger = logger.With("module", "main")
+
+	return SetCmdServerContext(cmd, serverCtx)
+}
+
+// GetServerContextFromCmd returns a Context from a command or an empty Context
+// if it has not been set.
+func GetServerContextFromCmd(cmd *cobra.Command) *Context {
+	if v := cmd.Context().Value(ServerContextKey); v != nil {
+		serverCtxPtr := v.(*Context)
+		return serverCtxPtr
+	}
 
-		logger = logger.With("module", "main")
-		ctx.Config = config
-		ctx.Logger = logger
+	return NewDefaultContext()
+}
 
-		return nil
+// SetCmdServerContext sets a command's Context value to the provided argument.
+func SetCmdServerContext(cmd *cobra.Command, serverCtx *Context) error {
+	v := cmd.Context().Value(ServerContextKey)
+	if v == nil {
+		return errors.New("server context not set")
 	}
+
+	serverCtxPtr := v.(*Context)
```

this will panic if the type is invalid, should we return an error instead?

---

No, it's guaranteed to be that type.